### ISSUE TITLE
Problems with afrodite response

### ISSUE THREAD

**Issue by vetlewi** [oslocyclotronlab/ompy#81]
I wanted to do some analysis with OMpy with the afrodite_labr2015 response matrix. I took the values for FE, SE, DE, etc. from the `folding.f` file in `MaMa` and created a response package (I've attached it to this issue). The resulting interpolation of the response matrix was this:
![resp](https://user-images.githubusercontent.com/5293376/69241501-48cb0300-0be2-11ea-8fb1-381c29f02488.png)
Clearly something is wrong here. I think that it might be related to the fact that only spectra up to 7 MeV is included and that is the reason for the fail beyond 7 MeV, though it does not explain some of the diagonals that's also strange.

**Comment by vetlewi** [comment]
Note that MaMa also fails above 7 MeV!
![resp_mama](https://user-images.githubusercontent.com/5293376/69241898-14a41200-0be3-11ea-9b99-958baecb8f24.png)

**Comment by vetlewi** [comment]
[afrodite_labr2015.zip](https://github.com/oslocyclotronlab/ompy/files/3869355/afrodite_labr2015.zip)

**Comment by vetlewi** [comment]
To clarify, I think the issue might be more related to the response function, rather than the interpolation... So I might have to figure this out myself...!

**Comment by fzeiser** [comment]
Can you dump the information from debug?

**Comment by fzeiser** [comment]
I have the geant flies if we want to have higher gamma energies

**Comment by vetlewi** [comment]
2019-11-21 01:02:16,761 - ompy.response - DEBUG - Loading response from file: afrodite_labr2015.zip
2019-11-21 01:02:17,146 - ompy.response - INFO - Note: Spectra outside of 100.0 and7000.0 are extrapolation only.
2019-11-21 01:02:17,643 - ompy.response - DEBUG - In fan method, 1 entries in Rare negative and now set to 0
2019-11-21 01:02:17,650 - ompy.response - DEBUG - In fan method, 1 entries in Rare negative and now set to 0
2019-11-21 01:02:17,658 - ompy.response - DEBUG - In fan method, 1 entries in Rare negative and now set to 0
2019-11-21 01:02:17,666 - ompy.response - DEBUG - In fan method, 3 entries in Rare negative and now set to 0
2019-11-21 01:02:17,674 - ompy.response - DEBUG - In fan method, 3 entries in Rare negative and now set to 0
2019-11-21 01:02:17,685 - ompy.response - DEBUG - In fan method, 5 entries in Rare negative and now set to 0
2019-11-21 01:02:17,693 - ompy.response - DEBUG - In fan method, 5 entries in Rare negative and now set to 0
2019-11-21 01:02:17,702 - ompy.response - DEBUG - In fan method, 7 entries in Rare negative and now set to 0
2019-11-21 01:02:17,711 - ompy.response - DEBUG - In fan method, 7 entries in Rare negative and now set to 0
2019-11-21 01:02:17,719 - ompy.response - DEBUG - In fan method, 7 entries in Rare negative and now set to 0
2019-11-21 01:02:17,728 - ompy.response - DEBUG - In fan method, 7 entries in Rare negative and now set to 0
2019-11-21 01:02:17,738 - ompy.response - DEBUG - In fan method, 10 entries in Rare negative and now set to 0
2019-11-21 01:02:17,747 - ompy.response - DEBUG - In fan method, 11 entries in Rare negative and now set to 0
2019-11-21 01:02:17,755 - ompy.response - DEBUG - In fan method, 11 entries in Rare negative and now set to 0
2019-11-21 01:02:17,763 - ompy.response - DEBUG - In fan method, 10 entries in Rare negative and now set to 0
2019-11-21 01:02:17,772 - ompy.response - DEBUG - In fan method, 15 entries in Rare negative and now set to 0
2019-11-21 01:02:17,781 - ompy.response - DEBUG - In fan method, 16 entries in Rare negative and now set to 0
2019-11-21 01:02:17,789 - ompy.response - DEBUG - In fan method, 21 entries in Rare negative and now set to 0
2019-11-21 01:02:17,797 - ompy.response - DEBUG - In fan method, 20 entries in Rare negative and now set to 0
2019-11-21 01:02:17,807 - ompy.response - DEBUG - In fan method, 20 entries in Rare negative and now set to 0
2019-11-21 01:02:17,818 - ompy.response - DEBUG - In fan method, 19 entries in Rare negative and now set to 0
2019-11-21 01:02:17,826 - ompy.response - DEBUG - In fan method, 18 entries in Rare negative and now set to 0
2019-11-21 01:02:17,834 - ompy.response - DEBUG - In fan method, 19 entries in Rare negative and now set to 0
2019-11-21 01:02:17,842 - ompy.response - DEBUG - In fan method, 23 entries in Rare negative and now set to 0
2019-11-21 01:02:17,850 - ompy.response - DEBUG - In fan method, 25 entries in Rare negative and now set to 0
2019-11-21 01:02:17,858 - ompy.response - DEBUG - In fan method, 23 entries in Rare negative and now set to 0
2019-11-21 01:02:17,869 - ompy.response - DEBUG - In fan method, 26 entries in Rare negative and now set to 0
2019-11-21 01:02:17,878 - ompy.response - DEBUG - In fan method, 28 entries in Rare negative and now set to 0
2019-11-21 01:02:17,886 - ompy.response - DEBUG - In fan method, 28 entries in Rare negative and now set to 0
2019-11-21 01:02:17,894 - ompy.response - DEBUG - In fan method, 31 entries in Rare negative and now set to 0
2019-11-21 01:02:17,902 - ompy.response - DEBUG - In fan method, 36 entries in Rare negative and now set to 0
2019-11-21 01:02:17,912 - ompy.response - DEBUG - In fan method, 30 entries in Rare negative and now set to 0
2019-11-21 01:02:17,920 - ompy.response - DEBUG - In fan method, 36 entries in Rare negative and now set to 0
2019-11-21 01:02:17,929 - ompy.response - DEBUG - In fan method, 34 entries in Rare negative and now set to 0
2019-11-21 01:02:17,937 - ompy.response - DEBUG - In fan method, 34 entries in Rare negative and now set to 0
2019-11-21 01:02:17,946 - ompy.response - DEBUG - In fan method, 34 entries in Rare negative and now set to 0
2019-11-21 01:02:17,954 - ompy.response - DEBUG - In fan method, 42 entries in Rare negative and now set to 0
2019-11-21 01:02:17,963 - ompy.response - DEBUG - In fan method, 37 entries in Rare negative and now set to 0
2019-11-21 01:02:17,971 - ompy.response - DEBUG - In fan method, 38 entries in Rare negative and now set to 0
2019-11-21 01:02:17,980 - ompy.response - DEBUG - In fan method, 39 entries in Rare negative and now set to 0
2019-11-21 01:02:17,988 - ompy.response - DEBUG - In fan method, 48 entries in Rare negative and now set to 0
2019-11-21 01:02:17,997 - ompy.response - DEBUG - In fan method, 39 entries in Rare negative and now set to 0
2019-11-21 01:02:18,006 - ompy.response - DEBUG - In fan method, 42 entries in Rare negative and now set to 0
2019-11-21 01:02:18,015 - ompy.response - DEBUG - In fan method, 47 entries in Rare negative and now set to 0
2019-11-21 01:02:18,024 - ompy.response - DEBUG - In fan method, 41 entries in Rare negative and now set to 0
2019-11-21 01:02:18,033 - ompy.response - DEBUG - In fan method, 43 entries in Rare negative and now set to 0
2019-11-21 01:02:18,043 - ompy.response - DEBUG - In fan method, 39 entries in Rare negative and now set to 0
2019-11-21 01:02:18,052 - ompy.response - DEBUG - In fan method, 44 entries in Rare negative and now set to 0
2019-11-21 01:02:18,061 - ompy.response - DEBUG - In fan method, 51 entries in Rare negative and now set to 0
2019-11-21 01:02:18,070 - ompy.response - DEBUG - In fan method, 52 entries in Rare negative and now set to 0
2019-11-21 01:02:18,080 - ompy.response - DEBUG - In fan method, 50 entries in Rare negative and now set to 0
2019-11-21 01:02:18,089 - ompy.response - DEBUG - In fan method, 50 entries in Rare negative and now set to 0
2019-11-21 01:02:18,099 - ompy.response - DEBUG - In fan method, 50 entries in Rare negative and now set to 0

**Comment by vetlewi** [comment]
Please send me the geant files, and I will add them to the response!

**Comment by fzeiser** [comment]
Ok, `info` at least spits out the "warning" / information. That's good.

But coming back to your issue: I believe that the response matrices there were created with the geant4 model we inherited/edited: https://github.com/oslocyclotronlab/AFRODITE

@grytveten : Can you confirm this? Do you still have the macros you used to run the simulation? If they are on *isotop*, I might just run a few simulations more. Otherwise I can set up something like we did for OSCAR. (+ anyhow link the spectrum collection the same way).

@vetlewi: Up to what energies do  you need the response?
BTW: By the geant files I mean the model, which is in the repo above; I don't have "ready simulrated" data on more. But I assume that we'll just simulate a few more spectra and process all simulration the same way as we did for OSCAR now.

We'll also need the resolution as a function of the energy.

**Comment by vetlewi** [comment]
> @vetlewi: Up to what energies do you need the response?
BTW: By the geant files I mean the model, which is in the repo above; I don't have "ready simulrated" data on more. But I assume that we'll just simulate a few more spectra and process all simulration the same way as we did for OSCAR now.

The resolution as a function of energy is the same as for OSCAR, so I can in principle use those. I think ideally the response matrix should have the same range as OSCAR. The highest energy that it is needed for is ~8 MeV. But don't start spending a lot of time on simulation just yet. The setup changed slightly between the 87Kr experiment and the 85Kr/133Xe experiment and the model would probably need to be modified slightly.

**Comment by grytveten** [comment]
Yes, I edited and simulated with some version of the AFRODITE setup of
Kevin. I did the simulations on my laptop, as I was traveling when I
finished that stuff. I'll try to have a look tonight.

**Comment by fzeiser** [comment]
So you want the simulations for the LaBr3's only I understand, right?

**Comment by fzeiser** [comment]
> Det er veldig hyggelig at dere tar dere tid til å se på dette, men det er virkelig ikke nødvendig å bruke mye tid. Med skrittene for å simulere dataene så får jeg til å gjøre simuleringene selv. Grunnen til at jeg lagde et issue på GitHub var først og fremst for å dokumentere at det var litt problemer med response funksjonen som må sees nærmere på, men er ikke noe som må løses ASAP.

I had problems in getting CADMesh compiled. @vetlewi Can you try to compile CADMesh and  then compile the AFRODITE (geant4) repo? Once you have done this, I'll get you set up on how to simulate the data and how to analyze it. We can convert the data to the new input format, so no hardcoding needed.

**Comment by vetlewi** [comment]
> > Det er veldig hyggelig at dere tar dere tid til å se på dette, men det er virkelig ikke nødvendig å bruke mye tid. Med skrittene for å simulere dataene så får jeg til å gjøre simuleringene selv. Grunnen til at jeg lagde et issue på GitHub var først og fremst for å dokumentere at det var litt problemer med response funksjonen som må sees nærmere på, men er ikke noe som må løses ASAP.
> 
> I had problems in getting CADMesh compiled. @vetlewi Can you try to compile CADMesh and then compile the AFRODITE (geant4) repo? Once you have done this, I'll get you set up on how to simulate the data and how to analyze it. We can convert the data to the new input format, so no hardcoding needed.

Yup. But won't be until next week... Or week after that :p

**Comment by vetlewi** [comment]
# Revisited!

## Fluctuation
The issue that causing large fluctuations of the response matrix are probably due to a poor smoothing of the input
<img width="577" alt="Skjermbilde 2020-03-26 kl  13 27 51" src="https://user-images.githubusercontent.com/5293376/77646845-9e833380-6f65-11ea-89d9-f4e62dff7586.png">
and will probably be easily fixed by redoing the smoothing.

## Problems above 7 MeV
The breakdown above 7 MeV is caused by the highest simulated energy being 7 MeV. Additional energies above 7 MeV should be ran in order to get better results.

**Comment by vetlewi** [comment]
> # Revisited!
> ## Fluctuation
> The issue that causing large fluctuations of the response matrix are probably due to a poor smoothing of the input
> <img alt="Skjermbilde 2020-03-26 kl 13 27 51" width="577" src="https://user-images.githubusercontent.com/5293376/77646845-9e833380-6f65-11ea-89d9-f4e62dff7586.png">
> and will probably be easily fixed by redoing the smoothing.

I've re-done the smoothing, with great results! Will repackage the response matrix and test!
<img width="576" alt="Skjermbilde 2020-03-26 kl  14 18 56" src="https://user-images.githubusercontent.com/5293376/77651318-e8bbe300-6f6c-11ea-8210-0552676a3091.png">

**Comment by fzeiser** [comment]
You may want to do the smoothing from within ompy. There it's easier to do a smoothing that is proportional to the fwhm with the `gauss_smoothing` function. 

Something along these lines:
``` py
def fFWHM(E, p):
    # FWHM in keV
    return np.sqrt(p[0] + p[1] * E + p[2]*E**2)

# Franks analysis, spring 2019
pars = np.array([73.2087, 0.50824, 9.62481e-05])
```

Then use `gauss_smoothing_matrix_1D` from ompy, where 
`fwhm =  fFWHM(matrix.Eg, *pars)`

**Comment by fzeiser** [comment]
Btw, thinking about it, I don't know whether Gry every smoothed these spectra!

**Comment by vetlewi** [comment]
> Btw, thinking about it, I don't know whether Gry every smoothed these spectra!

Seems like that might be the problem

**Comment by vetlewi** [comment]
> > # Revisited!
> > ## Fluctuation
> > The issue that causing large fluctuations of the response matrix are probably due to a poor smoothing of the input
> > <img alt="Skjermbilde 2020-03-26 kl 13 27 51" width="577" src="https://user-images.githubusercontent.com/5293376/77646845-9e833380-6f65-11ea-89d9-f4e62dff7586.png">
> > and will probably be easily fixed by redoing the smoothing.
> 
> I've re-done the smoothing, with great results! Will repackage the response matrix and test!
> <img alt="Skjermbilde 2020-03-26 kl 14 18 56" width="576" src="https://user-images.githubusercontent.com/5293376/77651318-e8bbe300-6f6c-11ea-8210-0552676a3091.png">

Response matrices after smoothing:
![ompy](https://user-images.githubusercontent.com/5293376/77653445-f9ba2380-6f6f-11ea-8b05-af64a14a8905.png)
![mama](https://user-images.githubusercontent.com/5293376/77653546-2110f080-6f70-11ea-8273-510016e09264.png)

I will smooth the matrices with OMpy as well to see if it has any impact, but I'm happy with the results :)

**Comment by vetlewi** [comment]
afrodite_labr2015.zip

**Comment by vetlewi** [comment]
And the response matrix after 'fixing':
[afrodite_labr2015.zip](https://github.com/oslocyclotronlab/ompy/files/4387481/afrodite_labr2015.zip)

**Comment by fzeiser** [comment]
Up to what energies do you use he unfolding? It seems like they are ok up to E_sim_max ≃ 7 MeV

**Comment by fzeiser** [comment]
Upload them to https://github.com/oslocyclotronlab/OCL_response_functions

**Comment by grytveten** [comment]
I don't remember anymore, but I think I maybe did not smooth.

**Comment by vetlewi** [comment]
> Please upload them to https://github.com/oslocyclotronlab/OCL_response_functions; "response matrices used at ocl, mostly oscar and cactus "

👍

**Comment by vetlewi** [comment]
I think that this is not really an issue anymore as the updated response matrix has been uploaded.
A small recap before I close the issue:
I experienced large fluctuations in the response matrix produced by OMpy. After further investigation I discovered that the same fluctuations was also present in the response matrix produced by MaMa. In the end the issue was due to insufficient smoothing of the Compton background spectra. As can be seen in this figure: <img alt="Skjermbilde 2020-03-26 kl 13 27 51" width="577" src="https://user-images.githubusercontent.com/5293376/77646845-9e833380-6f65-11ea-89d9-f4e62dff7586.png">
the Compton background spectra had significant fluctuations that had not been correctly smoothed. After smoothing the issue more or less disappeared.
